pkg/git/git.go

```diff
@@ -88,28 +89,59 @@ func NewGitClientFromExecutor(executor executor.Executor) *Client {
 	}
 }
 
-// CloneRepo clones a repo specified in the gitURL and checks out the specified branch
-func (git *Client) CloneRepo(cloneDirPrefix string, branch string, gitURL string) (string, error) {
-	cloneDir, err := ioutil.TempDir(os.TempDir(), cloneDirPrefix)
+// CloneOptions are the options for cloning a Git repository
+type CloneOptions struct {
+	URL       string
+	Branch    string
+	Bootstrap bool // create the branch if the repository is empty
+}
+
+// CloneRepoInTmpDir clones a repo specified in the gitURL in a temporary directory and checks out the specified branch
+func (git *Client) CloneRepoInTmpDir(tmpDirPrefix string, options CloneOptions) (string, error) {
+	cloneDir, err := ioutil.TempDir(os.TempDir(), tmpDirPrefix)
 	if err != nil {
 		return "", fmt.Errorf("cannot create temporary directory: %s", err)
 	}
-
-	return cloneDir, git.CloneRepoInPath(cloneDir, branch, gitURL)
+	return cloneDir, git.cloneRepoInPath(cloneDir, options)
 }
 
-// CloneRepoInPath clones a repo to the specified directory
-func (git *Client) CloneRepoInPath(clonePath string, branch string, gitURL string) error {
+// CloneRepoInPath behaves like CloneRepoInTmpDir but clones the repository in a specific directory
+// which creates if needed
+func (git *Client) CloneRepoInPath(clonePath string, options CloneOptions) error {
 	if err := os.MkdirAll(clonePath, 0700); err != nil {
 		return errors.Wrapf(err, "unable to create directory for cloning")
 	}
-	args := []string{"clone", "-b", branch, gitURL, clonePath}
+	return git.cloneRepoInPath(clonePath, options)
+}
+
+func (git *Client) cloneRepoInPath(clonePath string, options CloneOptions) error {
+	args := []string{"clone", options.URL, clonePath}
 	if err := git.runGitCmd(args...); err != nil {
 		return err
 	}
-	// Set the working directory only after the clone so that
-	// it doesn't create an undesirable nested directory
+	// Set the working directory to the cloned directory, but
+	// only do it after the clone so that it doesn't create an
+	// undesirable nested directory
 	git.dir = clonePath
+
+	if options.Branch != "" {
+		// Switch to target branch
+		args := []string{"checkout", options.Branch}
+		if options.Bootstrap {
+			// create the branch if the repository is empty (no pre-existing branches)
+			files, err := ioutil.ReadDir(filepath.Join(git.dir, ".git", "refs", "heads"))
```

Comment:
Nit: extract to utility function:
```go
func doBranchesExist(dir string) (bool, error) {
	files, err := ioutil.ReadDir(filepath.Join(dir, ".git", "refs", "heads"))
	if err != nil {
		return false, err
	}
	return len(files) != 0, nil
}
```
to separate the "what" from the "how", and hopefully increase readability.

Reply:
Good point. Will do